File: frontend/src/types/transaction.ts
export type TransactionType = 'INCOME' | 'EXPENSE' | 'TRANSFER'

export interface TransactionResponse {
  id: number
  accountId: number
  accountName: string
  categoryId: number | null
  categoryName: string | null
  categoryIcon: string | null
  categoryColor: string | null
  amount: number
  currency: string
  convertedAmount: number | null
  transactionType: TransactionType
  merchant: string | null
  description: string | null
  transactionDate: string
  isRecurring: boolean
  isAutoCategorized: boolean
  tags: string | null
  createdAt: string
}

export interface TransactionRequest {
  accountId: number
  categoryId?: number
  amount: number
  currency: string
  transactionType: TransactionType
  merchant?: string
  description?: string
  transactionDate: string
  toAccountId?: number
  tags?: string
}

export interface TransactionFilters {
  page?: number
  size?: number
  from?: string
  to?: string
  categoryId?: number
  accountId?: number
  type?: TransactionType
  search?: string
}

File: src/main/java/com/aswin/moneymanager/entity/Transaction.java
package com.aswin.moneymanager.entity;

import com.aswin.moneymanager.enums.TransactionType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "transactions")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Transaction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "account_id", nullable = false)
    private Account account;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal amount;

    @Column(nullable = false, length = 3)
    private String currency;

    @Column(name = "converted_amount", precision = 15, scale = 2)
    private BigDecimal convertedAmount;

    @Enumerated(EnumType.STRING)
    @Column(name = "transaction_type", nullable = false)
    private TransactionType transactionType;

    @Column(length = 200)
    private String merchant;

    @Column(length = 500)
    private String description;

    @Column(name = "transaction_date", nullable = false)
    private LocalDate transactionDate;

    @Column(name = "is_recurring", nullable = false)
    @Builder.Default
    private Boolean isRecurring = false;

    @Column(name = "recurring_group_id")
    private Long recurringGroupId;

    @Column(name = "is_auto_categorized", nullable = false)
    @Builder.Default
    private Boolean isAutoCategorized = false;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "tags")
    private String tags;

    @Column(name = "receipt_url", length = 500)
    private String receiptUrl;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}
